streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultStateConstrainedBalancedAssignor.java

```diff
@@ -0,0 +1,252 @@
+        assignTasksWithCaughtUpClients(assignment, tasksToCaughtUpClients, statefulTasksToRankedClients);
+        assignTasksWithoutCaughtUpClients(assignment, tasksToCaughtUpClients, statefulTasksToRankedClients);
+        balance(assignment, balanceFactor, statefulTasksToRankedClients, tasksToCaughtUpClients);
+        return assignment;
+    }
+
+    /**
+     * Initialises the assignment with the clients in the ranked clients map and assigns to each client an empty list.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @return initialised assignment with empty lists
+     */
+    private Map<ID, List<TaskId>> initAssignment(final SortedMap<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
+        final Map<ID, List<TaskId>> assignment = new HashMap<>();
+        final SortedSet<ClientIdAndLag<ID>> clientIdAndLags = statefulTasksToRankedClients.values().stream()
+            .findFirst().orElseThrow(() -> new IllegalStateException("The list of clients and lags must not be empty"));
+        for (final ClientIdAndLag<ID> clientIdAndLag : clientIdAndLags) {
+            assignment.put(clientIdAndLag.clientId(), new ArrayList<>());
+        }
+        if (assignment.isEmpty()) {
+            throw new IllegalStateException("Clients within an assignment must not be empty");
+        }
+        return assignment;
+    }
+
+    /**
+     * Maps tasks to clients with caught-up states for the task.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @return map from tasks with caught-up clients to the list of client candidates
+     */
+    private Map<TaskId, List<ID>> tasksToCaughtUpClients(final SortedMap<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
+        final Map<TaskId, List<ID>> taskToCaughtUpClients = new HashMap<>();
+        for (final SortedMap.Entry<TaskId, SortedSet<ClientIdAndLag<ID>>> taskToRankedClient : statefulTasksToRankedClients.entrySet()) {
+            final SortedSet<ClientIdAndLag<ID>> rankedClients = taskToRankedClient.getValue();
+            for (final ClientIdAndLag<ID> clientIdAndLag : rankedClients) {
+                if (clientIdAndLag.lag() == Task.LATEST_OFFSET || clientIdAndLag.lag() == 0) {
+                    final TaskId taskId = taskToRankedClient.getKey();
+                    taskToCaughtUpClients.computeIfAbsent(taskId, ignored -> new ArrayList<>()).add(clientIdAndLag.clientId());
+                } else {
+                    break;
+                }
+            }
+        }
+        return taskToCaughtUpClients;
+    }
+
+    /**
+     * Maps a task to the client that host the task according to the previous assignment.
+     *
+     * @return map from task IDs to clients hosting the corresponding task
+     */
+    private Map<TaskId, ID> tasksToPreviousClients(final Map<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
```

Comment:
req: clarify (either in name or in javadocs) that this only returns tasks whose previous client was caught-up (ie the task got to RUNNING)